### ISSUE TITLE
[Suggestion] Add after-sponsor buffering

### ISSUE THREAD

**Issue by playsenge** [ajayyy/SponsorBlock#543]
Hello!
While I love SponsorBlock, there's a minor issue/suggestion.
When I am watching a video with a sponsorship coming up at the beggining, when it skips after the sponsorship it hasn't buffered yet and has to load the part after the sponsorship. It's instead buffering the sponsorship (which is useless).

Hopefully this can get accepted!

Thanks,
senge.

**Comment by ajayyy** [comment]
Sadly, I do not think this is possible

**Comment by elypter** [comment]
it might be possible on an alternative youtube frontend that directly integrates sponsorblock. there are currently no plans that i know of for invidious or others to implement it though.

**Comment by playsenge** [comment]
Well, if they don't know about this or no one requests it to them they are not going to do it :P

**Comment by elypter** [comment]
it would be either your or the repositorys owner to ask for upstream changes. regardless i already asked 2 weeks ago and it was beyond the maintainers scope. so realisitcly you are only going to see this change if you do it yourself.

**Comment by NDevTK** [comment]
@playsenge I have been trying implement this it seem practical to implement.
Progress: https://gist.github.com/NDevTK/be7df78ac1b6e3dbf12ef97afa941c18

**Comment by NDevTK** [comment]
Sorry about the delay this issue has not been forgotten about.

**Comment by NDevTK** [comment]
tbh I did forget... may look into this more if I remember :)

**Comment by ccuser44** [comment]
The sponsor buffering should also not load the content inside the skippable sections to save bandwith. 

This could be done via directly modifying the yt loader javascriot

**Comment by NDevTK** [comment]
Seems like there would be compatibility issues with that.
If it went that far it should modify the video source so that skipping is instant.

**Comment by ajayyy** [comment]
Some notes from working on waveform editor implementation that might be helpful for this:

Hijacking `fetch` function will be able to take over the fetch request. Requests follow a pattern, and for after-sponsor buffering, playback URLs for future chunks could be pre-fetched from the m3u8 playlist and returned instantly when YouTube itself tries to fetch it to create an instant buffer effect.

`videoplayback` video requests all are almost the same except for a few fields that vary

![image](https://github.com/ajayyy/SponsorBlock/assets/12688112/b13d35e6-0611-41f5-87c9-a7dfa35dc6e8)

Ensuring the correct resolution is pulled would be important. Data should be accessible from ytInitialData or `player` requests made through fetch.

Fetch function is already being hijacked by document script for other reasons, so can reuse this injection point.

**Comment by NDevTK** [comment]
I would prefer something that's not hijacking network requests.
Maybe something like quickly changing the currentTime to end of sponsor then changing it back will make there player do the work.

**Comment by ccuser44** [comment]
> I would prefer something that's not hijacking network requests.
> Maybe something like quickly changing the currentTime to end of sponsor then changing it back will make there player do the work.

Changing the time back and forth is more hacky

**Comment by NDevTK** [comment]
CC: YouTube Product Team
What's the ETA on the extension custom buffer API?

It seems unlikely the browser cache will work well since URL changes would need to always be predictable.

If it was always an exact cache match you could just use the YouTube player API in a hidden iframe to preload times when needed.

**Comment by ajayyy** [comment]
I've done more research on DASH range requests and I don't see an easy way to predict what ranges for each chunk YouTube will use. It seems to vary, and there is no manifest file

**Comment by JxxIT** [comment]
@NDevTK Hi, any progress on this? Thanks for looking into this!

**Comment by NDevTK** [comment]
Yeah so this is still a thing... I don't use the extension and making this change would increase the risk of YouTube playback breaking. One fix that might work would be to get a faster internet connection if this is not practical in your area you have two options.

- Go someone else
- Use virtual desktop infrastructure (VDI) this makes it so the pointless buffering is handled by the remote server

Hope this helps, NDevTK Support.

**Comment by JxxIT** [comment]
I'm working on implementing this.
